src/coprocessor/dag/batch/executors/fast_hash_aggr_executor.rs

```diff
@@ -0,0 +1,544 @@
+            Arc::new(EvalConfig::default()),
+            src,
+            group_by_exp,
+            aggr_defs,
+            aggr_def_parser,
+        )
+        .unwrap()
+    }
+}
+
+impl BatchFastHashAggregationExecutor<ExecSummaryCollectorDisabled, Box<dyn BatchExecutor>> {
+    /// Checks whether this executor can be used.
+    #[inline]
+    pub fn check_supported(descriptor: &Aggregation) -> Result<()> {
+        let group_by_definitions = descriptor.get_group_by();
+        assert!(!group_by_definitions.is_empty());
+        if group_by_definitions.len() > 1 {
+            return Err(box_err!("Multi group is not supported"));
+        }
+
+        let def = &group_by_definitions[0];
+
+        // Only a subset of all eval types are supported.
+        let eval_type = box_try!(EvalType::try_from(def.get_field_type().tp()));
+        match eval_type {
+            EvalType::Int | EvalType::Real | EvalType::Bytes | EvalType::Duration => {}
+            _ => return Err(box_err!("Eval type {} is not supported", eval_type)),
+        }
+
+        RpnExpressionBuilder::check_expr_tree_supported(def)?;
+        if RpnExpressionBuilder::is_expr_eval_to_scalar(def)? {
+            return Err(box_err!("Group by expression is not a column"));
+        }
+
+        let aggr_definitions = descriptor.get_agg_func();
+        for def in aggr_definitions {
+            AllAggrDefinitionParser.check_supported(def)?;
+        }
+        Ok(())
+    }
+}
+
+impl<C: ExecSummaryCollector, Src: BatchExecutor> BatchFastHashAggregationExecutor<C, Src> {
+    pub fn new(
+        summary_collector: C,
+        config: Arc<EvalConfig>,
+        src: Src,
+        group_by_exp_defs: Vec<Expr>,
+        aggr_defs: Vec<Expr>,
+    ) -> Result<Self> {
+        assert_eq!(group_by_exp_defs.len(), 1);
+        let group_by_exp = RpnExpressionBuilder::build_from_expr_tree(
+            group_by_exp_defs.into_iter().next().unwrap(),
+            &config.tz,
+            src.schema().len(),
+        )?;
+        Self::new_impl(
+            summary_collector,
+            config,
+            src,
+            group_by_exp,
+            aggr_defs,
+            AllAggrDefinitionParser,
+        )
+    }
+
+    #[inline]
+    fn new_impl(
+        summary_collector: C,
+        config: Arc<EvalConfig>,
+        src: Src,
+        group_by_exp: RpnExpression,
+        aggr_defs: Vec<Expr>,
+        aggr_def_parser: impl AggrDefinitionParser,
+    ) -> Result<Self> {
+        let group_by_field_type = group_by_exp.ret_field_type(src.schema()).clone();
+        let group_by_eval_type = EvalType::try_from(group_by_field_type.tp()).unwrap();
+        let groups = match_template_hashable! {
+            TT, match group_by_eval_type {
+                EvalType::TT => Groups::TT(HashMap::default()),
+                _ => unreachable!(),
+            }
+        };
+
+        let aggr_impl = FastHashAggregationImpl {
+            states: Vec::with_capacity(1024),
```

Comment:
Hard code a `1024` here is not a good idea.

Reply:
We may never know what value is a good value for this one even in future 🤣

Reply:
I guess it is something that can be estimated 🤔  (Like cost estimation in TiDB?)

Reply:
Yes, TiDB can do this with the help of analyze data while we don't have them (and I think not very necessary because even for a precise value it will just save a few times of allocations).

Reply:
OK